lisp/core.slosh

```diff
@@ -655,6 +673,87 @@ Section: core
 #%
 Asserts the value is an error
 
-Section: core
+Section: test
 %#
 (defmacro test::assert-error (& body) `(assert-error ~@body))
+
+
+#%
+Usage: (apply function arg* list)
+
+Call the provided function with the supplied arguments, last is a list that will be expanded.
+
+Section: core
+
+Example:
+(def test-apply-one (apply str "O" "NE"))
+(test::assert-equal "ONE" test-apply-one)
+(test::assert-equal 10 (apply + 1 2 7))
+%#
+(defmacro apply (a-fn & rst)
+    `(~a-fn ~@rst))
```

Comment:
I couldn't get a version of apply to work that would return 6 for all of the following:

```slosh
(apply + 1 2 3)
(apply + (list 1 2 3))
(apply + 1 (list 2 3))
```
I could only get the first one working with this simple macro.

ideally it would work for all 3 right?

Reply:
This is probably an edge case, the + "function" is compiled so it will not behave as expected when used like this.  This is something I will need to fix, I started some of the support for this a while ago but it is not hooked up (and I have not totally decided how to do it yet).  Obviously you need to be able to use "functions" that are compiled (in this case for speed since this becomes a single op code) need to work as expected when used.